kotlinx-coroutines-core/nativeDarwin/test/MainDispatcherTest.kt

```diff
@@ -0,0 +1,146 @@
+            expect(1)
+            assertFalse(isMainThread())
+            withContext(Dispatchers.Main) {
+                assertTrue(isMainThread())
+                expect(2)
+            }
+            assertFalse(isMainThread())
+            finish(3)
+        }
+    }
+
+    @Test
+    fun testWithContextDelay() {
+        if (!canTestMainDispatcher()) return // skip if already on the main thread, run blocking doesn't really work well with that
+        runTest {
+            expect(1)
+            withContext(Dispatchers.Main) {
+                assertTrue(isMainThread())
+                expect(2)
+                delay(100)
+                assertTrue(isMainThread())
+                expect(3)
+            }
+            assertFalse(isMainThread())
+            finish(4)
+        }
+    }
+
+    @Test
+    fun testWithTimeoutContextDelayNoTimeout() {
+        if (!canTestMainDispatcher()) return // skip if already on the main thread, run blocking doesn't really work well with that
+        runTest {
+            expect(1)
+            withTimeout(1000) {
+                withContext(Dispatchers.Main) {
+                    assertTrue(isMainThread())
+                    expect(2)
+                    delay(100)
+                    assertTrue(isMainThread())
+                    expect(3)
+                }
+            }
+            assertFalse(isMainThread())
+            finish(4)
+        }
+    }
+
+    @Test
+    fun testWithTimeoutContextDelayTimeout() {
+        if (!canTestMainDispatcher()) return // skip if already on the main thread, run blocking doesn't really work well with that
+        runTest {
+            expect(1)
+             assertFailsWith<TimeoutCancellationException> {
+                withTimeout(100) {
+                    withContext(Dispatchers.Main) {
+                        assertTrue(isMainThread())
+                        expect(2)
+                        delay(1000)
+                        expectUnreached()
+                    }
+                }
+                expectUnreached()
+            }
+            assertFalse(isMainThread())
+            finish(3)
+        }
+    }
+
+    @Test
+    fun testWithContextTimeoutDelayNoTimeout() {
+        if (!canTestMainDispatcher()) return // skip if already on the main thread, run blocking doesn't really work well with that
+        runTest {
+            expect(1)
+            withContext(Dispatchers.Main) {
+                withTimeout(1000) {
+                    assertTrue(isMainThread())
+                    expect(2)
+                    delay(100)
+                    assertTrue(isMainThread())
+                    expect(3)
+                }
+            }
+            assertFalse(isMainThread())
+            finish(4)
+        }
+    }
+
+    @Test
+    fun testWithContextTimeoutDelayTimeout() {
+        if (!canTestMainDispatcher()) return // skip if already on the main thread, run blocking doesn't really work well with that
```

Comment:
Style: maybe this comment belongs to `canTestMainDispatcher` itself. Or even better: why not make a wrapper around `runTest` that does this check, like `runTestNotOnMainDispatcher`?